x/mongo/driver/auth/internal/gssapi/sspi_wrapper.c

```diff
@@ -9,7 +9,30 @@ static const LPSTR SSPI_PACKAGE_NAME = "kerberos";
 int sspi_init(
 )
 {
-	sspi_secur32_dll = LoadLibrary("secur32.dll");
+	// Load the secur32.dll library using its exact path. Passing the exact DLL path rather than allowing LoadLibrary to
+	// search in different locations removes the possibility of DLL preloading attacks.
+
+	// Passing a 0 size will return the required buffer length to hold the path, including the null terminator.
+	int requiredLen = GetSystemDirectory(NULL, 0);
+	if (!requiredLen) {
+		return GetLastError();
+	}
+
+	// Allocate a buffer to hold the system directory + "\secur32.dll" (length 12, not including null terminator).
+	int actualLen = requiredLen + 12;
+	char *directoryBuffer = (char *) calloc(1, actualLen);
```

Comment:
@markbenvenuto may need to correct me, but this may need to be:
```
(char *) calloc (sizeof (TCHAR), actualLen);
```

Judging from the note [here](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getsystemdirectoryw#examples) this switches between `GetSystemDirectoryW` or `GetSystemDirectoryA` depending on whether `UNICODE` is defined. [This page](https://docs.microsoft.com/en-us/windows/win32/intl/conventions-for-function-prototypes) indicates the allocation should be done in `TCHAR` since the size of `TCHAR` can differ:
>When a function has a length parameter for a character string, the length should be documented as a count of TCHAR values in the string. This data type refers to bytes for Windows code page versions of the function or 16-bit words for Unicode versions. However, functions that require or return pointers to untyped memory blocks, such as the GlobalAlloc function, generally take a size in bytes, regardless of the prototype that is used. If the allocation of untyped memory is for a string, the application must multiply the number of characters by sizeof(TCHAR). For additional information, see Using Generic Data Types.

Reply:
While @kevinAlbs is right about the variation in char sizes, I would prefer if we just switch to calling `GetSystemDirectoryA` and `LoadLibraryA` as the best way to address this. Now, this size variation relies on the assumption that a user endeavors to override the `cgo` compilation such that `UNICODE` is enabled. Today, it compiles the code is compiled as ASCII. A user would have to set environment flags or modify the code and I believe this to be a rather remote possibility.

Reply:
Changed to use GetSystemDirectoryA and LoadLibraryA. I'll keep an eye on the Evergreen patch to see if that causes any issues.

Reply:
Ok, thanks for the explanation!